common/lib/xmodule/xmodule/js/src/video/09_completion.js

```diff
@@ -0,0 +1,123 @@
+                this.endTime = this.state.config.endTime;
+                if (this.endTime) {
+                    this.completeAfter = this.startTime + (this.endTime - this.startTime) * this.completionPercentage;
+                }
+                this.bindHandlers();
+            },
+
+        /** Bind any necessary function callbacks to DOM events. */
+            bindHandlers: function() {
+                var self = this;
+
+            /** Event handler to check if the video is complete, and submit a completion if it is.
+             *
+             *  If the timeupdate handler doesn't fire after the required percentage, this will catch
+             *  any fully complete videos.
+             **/
+                this.state.el.on('ended.completion', function() {
+                    if (self.complete) {
+                        return;
+                    }
+                    self.markCompletion();
+                });
+
+            /** Event handler to check video progress, and mark complete if
+             *  greater than completionPercentage
+             **/
+                this.state.el.on('timeupdate.completion', function(ev, currentTime) {
+                    var duration;
+                    if (self.complete) {
+                        return;
+                    }
+                    if (self.lastSentAt !== undefined && currentTime - self.lastSentAt < 3.0) {
+                        // Throttle attempts to submit in case of network issues
+                        return;
+                    }
+                    if (self.completeAfter === undefined) {
+                        // Duration is not available at initialization time
+                        duration = self.state.videoPlayer.duration();
+                        if (duration === 0) {
+                            // Video is too short, or duration is not yet set.
+                            // Wait for another event, or fall back to
+                            // 'ended' handler.
+                            return;
+                        }
+                        self.completeAfter = self.startTime + (duration - self.startTime) * self.completionPercentage;
+                    }
+
+                    if (currentTime > self.completeAfter) {
+                        this.lastSentAt = currentTime;
+                        self.markCompletion();
+                    }
+                });
+            },
+
+            markCompletion: function() {
+                var self = this;
+                var response;
+                var errmsg;
+                this.complete = true;
+                if (this.state.config.publishCompletionUrl) {
+                    response = $.ajax({
+                        type: 'POST',
+                        url: this.state.config.publishCompletionUrl,
+                        contentType: 'application/json',
+                        data: JSON.stringify({completion: 1.0})
```

Comment:
Cleaner way to post JSON data in [035_video_accessible_menu.js](https://github.com/open-craft/edx-platform/blob/1b9342faaa0ca8c0c6f118e70cc56c0ee10f2699/common/lib/xmodule/xmodule/js/src/video/035_video_accessible_menu.js#L52-L63)

Reply:
When I try that (`data: {completion: 1.0}, dataType: 'json'`), it submits the data as x-www-form-encoded, and returns a 400 error.  I've changed it back, but left in a commit (https://github.com/edx/edx-platform/pull/16569/commits/a0f7ed57a14de2b30c539de9638801b65c571e20) that tries to do it the way 035_video_accessible_menu does it the way it's done in 035_video_accessible_menu.js, in case you can see something obvious I was missing.